### PR TITLE
Add web workflow to ESP boards

### PR DESCRIPTION
**Pull request by tannewt**
This adds a web server to CircuitPython that runs on port 80 and provides workflow related APIs. See `docs/workflows.md` for details.

Related to  #6174

### CODE REVIEW

**Review by dhalbert — COMMENTED** [review]
This is great!

I tried giving an incorrect password again. When connected to the REPL, I see a long delay before the REPL produces output, and even after producing output, I cannot type at the REPL for 10-15 seconds after "Press any key" shows up.

If I have two boards web-capable boards, at least on Linux, and unplug one, and plug in the other, the avahi caching will resolve `circuitpython.local` to the old board's IP address for some time (minutes, it appears). I don't know how well it works on Windows or macOS. There doesn't seem to be an easy way to clear the cache except to restart the avahi daemon.

Rather mysteriously, the first board I tried did not show up in the browser for some time. This seems like a DNS issue.

I could not get the Javacript buttons to work in the file browser -- they were all gray all the time.

I think we could still merge this soon for people to try, and say it's rough and we want feedback.

File: supervisor/shared/web_workflow/web_workflow.c [hunk 1]
```diff
@@ -91,9 +226,935 @@ void supervisor_start_web_workflow(void) {
...
+    #if CIRCUITPY_USB && CIRCUITPY_USB_MSC
+    if (storage_usb_enabled() && !usb_msc_lock()) {
+        return true;
+    }
+    #endif
+    return false;
+}
+
+static void _cors_header(socketpool_socket_obj_t *socket, _request *request) {
+    _send_str(socket, "Access-Control-Allow-Credentials: true\r\nVary: Origin\r\n");
```
[the remark this answers is not in the document]
Comment (dhalbert) [reply supervisor/shared/web_workflow/web_workflow.c:352]: Maybe use the printf mechanism.

**Comment by dhalbert** [comment]
I'm thinking about a `CIRCUITPY_WEB_API_ENABLE` in `.env`. By default it would be off, so that you have set it to enable the web workflow to have it be on. This allows `.env` to store SSID and password as an eventual replacement for `secrets.py`. I can open this up as a separate issue later.

It also brings up the question of how to represent boolean values in `.env`.

**Comment by tannewt** [comment]
> I tried giving an incorrect password again. When connected to the REPL, I see a long delay before the REPL produces output, and even after producing output, I cannot type at the REPL for 10-15 seconds after "Press any key" shows up.

This should be the same as before. I didn't change anything with the auto-connect.

> If I have two boards web-capable boards, at least on Linux, and unplug one, and plug in the other, the avahi caching will resolve `circuitpython.local` to the old board's IP address for some time (minutes, it appears). I don't know how well it works on Windows or macOS. There doesn't seem to be an easy way to clear the cache except to restart the avahi daemon.

I haven't noticed this on Arch. Though I've been mainly using the same board. You should be able to enter the others IP and it'll do the MDNS search for you (and use the others' IPs in the list.)

> Rather mysteriously, the first board I tried did not show up in the browser for some time. This seems like a DNS issue.

MDNS uses UDP so it can take some time.

> I could not get the Javacript buttons to work in the file browser -- they were all gray all the time.

This is likely because you had USB storage enabled. I'll add a notice.

> I think we could still merge this soon for people to try, and say it's rough and we want feedback.

100%. puts the alpha in alpha. :-)

**Comment by tannewt** [comment]
> I'm thinking about a `CIRCUITPY_WEB_API_ENABLE` in `.env`. By default it would be off, so that you have set it to enable the web workflow to have it be on. This allows `.env` to store SSID and password as an eventual replacement for `secrets.py`. I can open this up as a separate issue later.

I'm open to doing this though I'd prefer it default on when using the auto-wifi. Folks can use other env variables if they want to control it themselves (maybe `HOME_WIFI` and `HOME_WIFI_PASSWORD` or `PORTAL_WIFI_SSID` and `PORTAL_WIFI_PASSWORD` for the portal library). `CIRCUITPY_` prefixed variables should be reserved for internal use.

> It also brings up the question of how to represent boolean values in `.env`.

`.env` doesn't have the concept of anything but a string: https://github.com/theskumar/python-dotenv/pull/293#issuecomment-782861756= Probably better to think of all values as enums instead of booleans. enums are also expandable which is nice. For webserver enable, maybe we want a key that is a comma separated list of ok ssids. That would turn it off if user code joins a different network.

**Comment by tannewt** [comment]
Ok, @dhalbert please take another look when you have time.

**Review by dhalbert — APPROVED** [review]
Thanks for the fixes. Looking forward to feedback.

**Pull request merged**